minetest/minetest
# Add crosshair Lua HUD element and replace hardcoded crosshair

## Goal of the PR
- Solves the crosshair part of  #9270
- With this PR adding new crosshair features also becomes much easier, for example adding item/node specific crosshairs or adding a crosshair style to solve #11889.

## What does this PR do?
- Adds a new Lua HUD element for crosshairs.
- Replaced the hardcoded crosshair by such a HUD element, such that it is only hardcoded in builtin.
- Makes the crosshair HUD flag only managed in builtin.
  (But still needs a `draw_crosshairs` condition since  `CAMERA_MODE_THIRD_FRONT` and `HAVE_TOUCHSCREENGUI` may disallow crosshairs. In the future it may be good to add a feature for Lua HudElements to specify CAMERA_MODE and TOUCHSCREEN behavior.)
- Adds a new style type for the crosshair HUD element, which makes it possible to dynamically set the crosshair textures. 
- Unscrambles the code for drawing HUD elements of type `HUD_ELEM_IMAGE`.
- Adds some HUD test code to devtest.

## Does this relate to a goal in [the roadmap](https://github.com/minetest/minetest/blob/master/doc/direction.md)?

- 2.3 UI Improvements
- 2.2 Internal code refactoring
- maybe also 2.1 Rendering/Graphics improvements

## To do

This PR is Ready for Review.

(The HUD builtin code (statbars.lua) should be refactored, but it is better to to this in a separate PR.)

## How to test
- Start devtest  and use the commands:
  - `/hudprint` to see that the predefined crosshair is now a Lua HUD element
  - `/hudcrosshairs` to show different crosshairs and check if the element parameters work as expected.
  - Point at an object to see whether the crosshairs change.
  - `/hudtoggleflag crosshair` to see if the flag disables the builtin crosshair.
  - `/hudimages` to check if the image HUD images still work.
- Use this to see if the builtin crosshair can be changed:
```lua
 minetest.hud_replace_builtin("crosshair", {
	type = "crosshair",
	position = {x=0.5, y=0.5},
	scale = {x = 10, y = 10},
})
```
- Execute `cp testhud_crosshair.png crosshair.png ; cp testhud_object_crosshair.png object_crosshair.png` inside `minetest/games/devtest/mods/testhud/textures ` to see if you can still use textures for the default crosshair.

## Screenshots
The crosshairs all use different parameters in the HUD element definition and on the left hand side you can see two image crosshairs (which are aligned). The crosshairs automatically change when you point at an object.
![Bildschirmfoto_2024-01-18_15-00-30](https://github.com/minetest/minetest/assets/7302071/5d4b8686-2310-42e5-83cb-b3c8be8648c5)
![Bildschirmfoto_2024-01-18_14-58-14](https://github.com/minetest/minetest/assets/7302071/f47a5d9a-388c-4ef6-928d-4c3efb6d708b)
